selfdrive/boardd/boardd.cc

```diff
@@ -644,48 +323,165 @@ void *can_recv_thread(void *crap) {
+
+    if (spoofing_started) {
+      health.ignition_line = 1;
+    }
+
+    voltage_f = VOLTAGE_K * (health.voltage / 1000.0) + (1.0 - VOLTAGE_K) * voltage_f;  // LPF
+
+    // Make sure CAN buses are live: safety_setter_thread does not work if Panda CAN are silent and there is only one other CAN node
+    if (health.safety_model == (uint8_t)(cereal::CarParams::SafetyModel::SILENT)) {
+      panda->set_safety_model(cereal::CarParams::SafetyModel::NO_OUTPUT);
+    }
+
+    bool ignition = ((health.ignition_line != 0) || (health.ignition_can != 0));
+
+    if (ignition) {
+      no_ignition_cnt = 0;
+    } else {
+      no_ignition_cnt += 1;
+    }
+
+  #ifndef __x86_64__
+    bool cdp_mode = health.usb_power_mode == (uint8_t)(cereal::HealthData::UsbPowerMode::CDP);
+    bool no_ignition_exp = no_ignition_cnt > NO_IGNITION_CNT_MAX;
+    if ((no_ignition_exp || (voltage_f < VBATT_PAUSE_CHARGING)) && cdp_mode && !ignition) {
+      std::vector<char> disable_power_down = read_db_bytes("DisablePowerDown");
+      if (disable_power_down.size() != 1 || disable_power_down[0] != '1') {
+        LOGW("TURN OFF CHARGING!\n");
+        panda->set_usb_power_mode(cereal::HealthData::UsbPowerMode::CLIENT);
+        LOGW("POWER DOWN DEVICE\n");
+        system("service call power 17 i32 0 i32 1");
+      }
+    }
+    if (!no_ignition_exp && (voltage_f > VBATT_START_CHARGING) && !cdp_mode) {
+      LOGW("TURN ON CHARGING!\n");
+      panda->set_usb_power_mode(cereal::HealthData::UsbPowerMode::CDP);
+    }
+
+    bool power_save_desired = !ignition;
+    if (health.power_save_enabled != power_save_desired){
+      panda->set_power_saving(power_save_desired);
+    }
+
+    // set safety mode to NO_OUTPUT when car is off. ELM327 is an alternative if we want to leverage athenad/connect
+    if (!ignition && (health.safety_model != (uint8_t)(cereal::CarParams::SafetyModel::NO_OUTPUT))) {
+      panda->set_safety_model(cereal::CarParams::SafetyModel::NO_OUTPUT);
+    }
+  #endif
+
+    // clear VIN, CarParams, and set new safety on car start
+    if (ignition && !ignition_last) {
+      int result = delete_db_value("CarVin");
+      assert((result == 0) || (result == ERR_NO_VALUE));
+      result = delete_db_value("CarParams");
+      assert((result == 0) || (result == ERR_NO_VALUE));
+
+      if (!safety_setter_thread_running) {
+        safety_setter_thread_running = true;
+        std::thread(safety_setter_thread).detach();
+      } else {
+        LOGW("Safety setter thread already running");
+      }
+    }
+
+    // Write to rtc once per minute when no ignition present
+    if ((panda->has_rtc) && !ignition && (no_ignition_cnt % 120 == 1)){
+      // Write time to RTC if it looks reasonable
+      struct tm sys_time = get_time();
+      if (time_valid(sys_time)){
+        panda->set_rtc(sys_time);
+      }
+    }
+
+    ignition_last = ignition;
+    uint16_t fan_speed_rpm = panda->get_fan_speed();
+
+    // set fields
```

Comment:
build health packet in panda class like the python library?

Reply:
actually not sure about this once since it'll create a dependency on cereal. python lib gets away with it because pycapnp can take a python dict

Reply:
there is already a dependency on cereal. But the health packet actually contains some other logic as well. It needs the ignition spoofing and the fan speed needs to be put in.